lib/script-generator.js

```diff
@@ -0,0 +1,209 @@
+'use strict';
+
+var FsDriver = require('./fs-driver');
+
+/**
+ * Generates shell scripts from transformation rules.
+ * @module fs-transform:script
+ * @author Ryan Sandor Richards
+ */
+module.exports = ScriptGenerator;
+
+/**
+ * Creates a new ScriptGenerator class that can generate scripts for the given
+ * set of transformation rules. Note: rules are not validated before generating
+ * the script. Please ensure you validate all rules using `Transformer.dry`
+ * before generating shell scripts.
+ * @class
+ * @param {array} rules Rules to convert into a shell script.
+ */
+function ScriptGenerator(rules) {
+  this.setRules(rules);
+  this.actionHandlers = {
+    'copy': this.copy.bind(this),
+    'rename': this.rename.bind(this),
+    'replace': this.replace.bind(this),
+    'exclude': this.exclude.bind(this)
+  };
+}
+
+/**
+ * Sets the rules for the script generator.
+ * @param {array} rules Rules to set for the generator.
+ */
+ScriptGenerator.prototype.setRules = function (rules) {
+  this.rules = rules;
+};
+
+/**
+ * Generates a shell script from the rules.
+ * @return {string} A shell script generated from the provided rules.
+ */
+ScriptGenerator.prototype.generate = function() {
+  return [
+    this.preamble(),
+    this.rules.map(this.generateRule.bind(this)).join('\n')
+  ].join('\n');
+};
+
+/**
+ * Generates the preamble for the shell script.
+ * @return {string} The preamble for the script.
+ */
+ScriptGenerator.prototype.preamble = function () {
+  var header = [
+    '#!/bin/sh\n',
+    '#',
+    '# Warning: this is a generated file, modifications may be overwritten.',
+    '#\n'
+  ].join('\n');
+
+  var common = [
+    '_script_name=`basename $0`',
+    'search_files=\'.\'',
+    'warning() { echo "($_script_name) WARNING" $1; }',
+    'error() { echo "($_script_name) ERROR" $1; exit 1; }\n'
+  ].join('\n');
+
+  var commandChecks = [
+    FsDriver.commands.map(function (name) {
+      return 'command -v ' + name + ' >/dev/null 2>&1 || {\n' +
+        '  error "Missing required command: ' + name + '";\n' +
+        '}';
+    }).join('\n'),
+  ].join('\n') + '\n';
+
+  return [header, common, commandChecks].join('\n');
+};
+
+/**
+ * Generates the script for the given rule.
+ * @param {object} rule Rule for which to generate the script.
+ * @return {string} The script that handles the given rule.
+ */
+ScriptGenerator.prototype.generateRule = function (rule, index) {
+  return this.actionHandlers[rule.action](rule, index);
+};
```

Comment:
1 - you're missing docs for `index`
2 - why is `index` required from the user? why isn't this something internal?